File: jd_logistics/backend/internal/driver/driver_model.go
package driver

import "jd_logistics/utils"

// ── Driver Profile ─────────────────────────────────────────────────────────────

type DriverProfile struct {
	utils.Model
	UserID          uint    `gorm:"uniqueIndex;not null" json:"user_id"`
	IsOnline        bool    `gorm:"default:false" json:"is_online"`
	IsVerified      bool    `gorm:"default:false" json:"is_verified"`
	Rating          float64 `gorm:"default:0" json:"rating"`
	TotalDeliveries int     `gorm:"default:0" json:"total_deliveries"`
	TotalEarnings   float64 `gorm:"default:0" json:"total_earnings"`
	VehicleType     string  `json:"vehicle_type"`
	VehicleNumber   string  `json:"vehicle_number"`
	LicenseNumber   string  `json:"license_number"`
	BankAccount     string  `json:"bank_account"`
	IFSC            string  `json:"ifsc"`
	CurrentLat      float64 `json:"current_lat"`
	CurrentLng      float64 `json:"current_lng"`
}

func (DriverProfile) TableName() string { return "jd_logistics.driver_profiles" }

// ── Vehicle ────────────────────────────────────────────────────────────────────

type Vehicle struct {
	utils.Model
	VehicleTypeID      *uint   `gorm:"index" json:"vehicle_type_id"`
	RegistrationNumber string  `gorm:"uniqueIndex;not null" json:"registration_number"`
	DriverID           *uint   `gorm:"index" json:"driver_id"`
	Make               string  `json:"make"`
	ModelName          string  `json:"model"`
	Year               int     `json:"year"`
	Color              string  `json:"color"`
	CapacityKg         float64 `json:"capacity_kg"`
	VolumeCbm          float64 `json:"volume_cbm"`
	FuelType           string  `json:"fuel_type"`
	InsuranceExpiry    *string `json:"insurance_expiry"`
	FitnessExpiry      *string `json:"fitness_expiry"`
	LastServiceDate    *string `json:"last_service_date"`
	IsActive           bool    `gorm:"default:true" json:"is_active"`
	CurrentLat         float64 `json:"current_lat"`
	CurrentLng         float64 `json:"current_lng"`
}

func (Vehicle) TableName() string { return "jd_logistics.vehicles" }

// ── Driver Wallet ──────────────────────────────────────────────────────────────

type DriverWallet struct {
	utils.Model
	DriverID       uint    `gorm:"uniqueIndex;not null" json:"driver_id"`
	Balance        float64 `gorm:"default:0" json:"balance"`
	PendingBalance float64 `gorm:"default:0" json:"pending_balance"`
	TotalEarned    float64 `gorm:"default:0" json:"total_earned"`
	TotalWithdrawn float64 `gorm:"default:0" json:"total_withdrawn"`
}

func (DriverWallet) TableName() string { return "jd_logistics.driver_wallet" }

// ── Earning Record ─────────────────────────────────────────────────────────────

type EarningRecord struct {
	utils.Model
	DriverID      uint    `gorm:"not null;index" json:"driver_id"`
	ShipmentID    uint    `gorm:"index" json:"shipment_id"`
	OrderID       *uint   `gorm:"index" json:"order_id"`
	OrderType     string  `gorm:"default:'courier'" json:"order_type"`
	Amount        float64 `json:"amount"`
	CommissionPct float64 `json:"commission_pct"`
	NetAmount     float64 `json:"net_amount"`
	Status        string  `gorm:"default:'pending'" json:"status"`
	Date          string  `json:"date"`
	PaidAt        *string `json:"paid_at"`
}

func (EarningRecord) TableName() string { return "jd_logistics.driver_earnings" }

// ── DTOs ───────────────────────────────────────────────────────────────────────

type UpdateLocationRequest struct {
	Latitude  float64 `json:"latitude" binding:"required"`
	Longitude float64 `json:"longitude" binding:"required"`
}

type ToggleOnlineRequest struct {
	IsOnline bool `json:"is_online"`
}

File: jd_logistics/frontend/lib/models/driver_model.dart
class DriverModel {
  final String id;
  final String userId;
  final bool isOnline;
  final bool isVerified;
  final double rating;
  final int totalDeliveries;
  final double totalEarnings;
  final String? vehicleType;
  final String? vehicleNumber;
  final String? licenseNumber;
  final double? currentLat;
  final double? currentLng;

  const DriverModel({
    required this.id,
    required this.userId,
    this.isOnline = false,
    this.isVerified = false,
    this.rating = 0.0,
    this.totalDeliveries = 0,
    this.totalEarnings = 0.0,
    this.vehicleType,
    this.vehicleNumber,
    this.licenseNumber,
    this.currentLat,
    this.currentLng,
  });

  factory DriverModel.fromJson(Map<String, dynamic> json) => DriverModel(
        id: json['id'].toString(),
        userId: json['user_id'].toString(),
        isOnline: json['is_online'] as bool? ?? false,
        isVerified: json['is_verified'] as bool? ?? false,
        rating: (json['rating'] as num? ?? 0).toDouble(),
        totalDeliveries: json['total_deliveries'] as int? ?? 0,
        totalEarnings: (json['total_earnings'] as num? ?? 0).toDouble(),
        vehicleType: json['vehicle_type'] as String?,
        vehicleNumber: json['vehicle_number'] as String?,
        licenseNumber: json['license_number'] as String?,
        currentLat: json['current_lat'] != null
            ? (json['current_lat'] as num).toDouble()
            : null,
        currentLng: json['current_lng'] != null
            ? (json['current_lng'] as num).toDouble()
            : null,
      );

  DriverModel copyWith({bool? isOnline}) => DriverModel(
        id: id,
        userId: userId,
        isOnline: isOnline ?? this.isOnline,
        isVerified: isVerified,
        rating: rating,
        totalDeliveries: totalDeliveries,
        totalEarnings: totalEarnings,
        vehicleType: vehicleType,
        vehicleNumber: vehicleNumber,
        licenseNumber: licenseNumber,
        currentLat: currentLat,
        currentLng: currentLng,
      );
}

class EarningModel {
  final String id;
  final String? shipmentId;
  final double amount;
  final String type;
  final String? description;
  final DateTime createdAt;

  const EarningModel({
    required this.id,
    this.shipmentId,
    required this.amount,
    required this.type,
    this.description,
    required this.createdAt,
  });

  factory EarningModel.fromJson(Map<String, dynamic> json) => EarningModel(
        id: json['id'].toString(),
        shipmentId: json['shipment_id']?.toString(),
        amount: (json['amount'] as num? ?? 0).toDouble(),
        type: json['type'] as String? ?? 'debit',
        description: json['description'] as String?,
        createdAt: json['created_at'] != null
            ? DateTime.tryParse(json['created_at'] as String) ?? DateTime.now()
            : DateTime.now(),
      );
}
